gluefactory/models/extractors/xfeat.py

```diff
@@ -0,0 +1,398 @@
+        rel_map = net_output['reliability_map']
+
+        B, _, H, W = x.shape
+
+        desc_map = F.normalize(desc_map, dim=1)
+
+        #Convert logits to heatmap and extract kpts
+        kpt_score_map = self.get_kpts_heatmap(kpt_logit_map)
+        mkpts = self.NMS(kpt_score_map)
+
+        #Compute reliability scores
+        _nearest = InterpolateSparse2d('nearest')
+        _bilinear = InterpolateSparse2d('bilinear')
+
+        scores = (_nearest(kpt_score_map, mkpts, H, W) * _bilinear(rel_map, mkpts, H, W)).squeeze(-1)
+        
+        # Filter out keypoints that are (0,0)
+        scores[torch.all(mkpts == 0, dim=-1)] = -1
+
+        #Select top-k features
+        idxs = torch.argsort(-scores)
+        mkpts_x  = torch.gather(mkpts[...,0], -1, idxs)[:, :self.top_k]
+        mkpts_y  = torch.gather(mkpts[...,1], -1, idxs)[:, :self.top_k]
+        mkpts = torch.cat([mkpts_x[...,None], mkpts_y[...,None]], dim=-1)
+        scores = torch.gather(scores, -1, idxs)[:, :self.top_k]
+
+        #Interpolate descriptors at kpts positions
+        feats = self.interpolator(desc_map, mkpts, H = H, W = W)
+
+        #L2-Normalize
+        feats = F.normalize(feats, dim=-1)
+
+        # Bring keypoints to original resolution
+        if self.conf.preprocess:
+            mkpts = mkpts * torch.tensor(
+                    [rw,rh], device=mkpts.device).view(1, 1, -1)
+
+        valid = scores > 0
+
+        keypoints_list = []
+        scores_list = []
+        descriptors_list = []
+    
+        if not self.conf.force_num_keypoints:
+            for b in range(B):
+                keypoints_list.append(mkpts[b][valid[b]])
+                scores_list.append(scores[b][valid[b]])
+                descriptors_list.append(feats[b][valid[b]])
+            
+            return {
+                "keypoints": keypoints_list,
+                "keypoint_scores": scores_list,
+                "descriptors": descriptors_list,
+            }
+        else:
+            return {
+                "keypoints": mkpts,
+                "keypoint_scores": scores,
+                "descriptors": feats,
+            }
+
+    def _forward(self, data: dict) -> dict:
+        """
+            input:
+                data -> Dict: contains 'image' key with torch.Tensor(B, C, H, W) grayscale or rgb images
+            return:
+                Dict: contains 'dense_descriptors', 'keypoint_logit_map', 'reliability_map' keys
+                      additionally contains 'keypoints', 'scores', 'descriptors' keys if get_sparse_outputs is True
+        """
+        image = data['image']
+
+        rh, rw = 1.0, 1.0
+        if self.conf.preprocess:
+            image, rh, rw = self.preprocess_tensor(image)
+
+        pred = self.net(image)
+
+        if self.conf.get_sparse_outputs:
+            pred_sparse = self.get_sparse_outputs(data, pred, rh, rw)
+            pred = {**pred, **pred_sparse}
+        
+        return pred
```

Comment:
Could you make sure we return the same keys (names and content) as with other point methods? i.e. keypoints, keypoints_scores and descriptors

Reply:
You can check it at Line 363

Reply:
Yeah, I saw it now :D sorry, my bad